### ISSUE TITLE
Multiple EXC_BAD_ACCESS crashes

### ISSUE THREAD

**Issue by iurichiba** [QuickBlox/quickblox-ios-sdk#660]
**Environment details**
iOS 8-10 with the following Pods (from .lock):
```
- QMChatViewController (0.3.10):
    - QMChatViewController/Categories (= 0.3.10)
    - QMChatViewController/KeyboardViewController (= 0.3.10)
    - QMChatViewController/Protocols (= 0.3.10)
    - QMChatViewController/Sections (= 0.3.10)
    - QMChatViewController/Utils (= 0.3.10)
    - QMChatViewController/Views (= 0.3.10)
    - QuickBlox (>= 2.0)
    - SDWebImage (~> 3.6)
    - TTTAttributedLabel (> 1.13)
  - QMChatViewController/Categories (0.3.10):
    - QMChatViewController/Utils
    - QuickBlox (>= 2.0)
    - SDWebImage (~> 3.6)
    - TTTAttributedLabel (> 1.13)
  - QMChatViewController/KeyboardViewController (0.3.10):
    - QuickBlox (>= 2.0)
    - SDWebImage (~> 3.6)
    - TTTAttributedLabel (> 1.13)
  - QMChatViewController/Protocols (0.3.10):
    - QuickBlox (>= 2.0)
    - SDWebImage (~> 3.6)
    - TTTAttributedLabel (> 1.13)
  - QMChatViewController/Sections (0.3.10):
    - QuickBlox (>= 2.0)
    - SDWebImage (~> 3.6)
    - TTTAttributedLabel (> 1.13)
  - QMChatViewController/Utils (0.3.10):
    - QuickBlox (>= 2.0)
    - SDWebImage (~> 3.6)
    - TTTAttributedLabel (> 1.13)
  - QMChatViewController/Views (0.3.10):
    - QMChatViewController/Categories
    - QMChatViewController/Protocols
    - QuickBlox (>= 2.0)
    - SDWebImage (~> 3.6)
    - TTTAttributedLabel (> 1.13)
  - QMServices (0.4.5):
    - Bolts (>= 1.5.0)
    - QuickBlox (>= 2.6.5)
  - QuickBlox (2.9)
```

**Did this work before?**
Tricky question, haha. I don't think so, but it's complicated:
- We can't replicate this crash, our tests couldn't pinpoint when and why it happens, but according to Crashlytics/Fabric, it does happen -- and it happens a lot: our crash-free users has plummeted from around 98% to around 90% since we implemented QB;
- Before the update, we used to get this same crash with the same stacktrace, but ending in some 3rd party library (KISSMetrics, Firebase, GoogleMaps). Now it ends within QB classes. I'm gonna attach Crashlytics logs, and if you want I can attach previous errors too (they're practically identical).

**Expected behavior**
Don't crash.

**Actual behavior**
It crashes.

**Logs**
- [Most frequent (block_invoke, EXC_BAD_ACCESS): 93 crashes, 72 users, in 4 days](http://imgur.com/a/bKds5)
- [Second-most frequent (block_invoke, EXC_BAD_ACCESS): 36 crashes, 33 users, in 4 days](http://imgur.com/a/nEiyt)

**Steps to reproduce the behavior**
Sorry, I don't know how. It'd probably be kind of a hit-and-miss scenario to fix it, but we hope that since you guys wrote that code, you can tell us what's happening.

**Any others comments?**
As commented above, it used to happen as a conflict with other libraries -- so we removed them. First we removed KISSMetrics, but then it started to happen in Firebase, then it started to happen in GoogleMaps, and so on. Then we finally thought that it was somehow QMService's or QB's fault, so we tried fiddling with your code, in the lines described to us by Crashlytics, but it didn't work.

It seems to always happen when calling a method w/ a block parameter. EXC_BAD_ACCESS tells us that it tried to reach something in memory and it wasn't there anymore (the target block?). We **think** it may have something to do with how long it takes to call those blocks. 

Also, we see lots of QB_warnBlockingOperationOnMainThread happening in these stacks, any idea what's that? (it starts with QB, so we're assuming it's related)

We didn't have much data to report it, but we're kinda desperate as we approach 2017. We **must** have it fixed before December. Now that the crash happens in QB classes, we can at least show you without doubt that for some reason it happens because of QB -- hope you can help us.

**Comment by Pro100Andrey** [comment]
Hi @iurichiba, Please provide code snippet for:
<img width="717" alt="crash1" src="https://cloud.githubusercontent.com/assets/1109916/20359149/3e11f42c-ac37-11e6-923d-43501b592c4d.png">

**Comment by iurichiba** [comment]
@Pro100Andrey here you go:

```obj-c
- (void)loadRecentDialogsWithSuccess:(void(^)(NSArray *dialogObjects))completion
                          andFailure:(void(^)(QBResponse *response))failure{
    NSDictionary *request = [self requestForRecentDialogs];
    [[ServicesManager instance].chatService allDialogsWithPageLimit:999 extendedRequest:request iterationBlock:^(QBResponse *response, NSArray *dialogObjects, NSSet *dialogsUsersIDs, BOOL *stop) {
        self.recentDialogs = self.dialogs.mutableCopy;
        if (completion != nil) completion(dialogObjects);
    } completion:^(QBResponse *response) {
        if ([ServicesManager instance].isAuthorized) {
            if (response.success) [ServicesManager instance].lastActivityDate = [NSDate date];
        } if (failure != nil) failure(response);
    }];
}

- (NSDictionary *)requestForRecentDialogs {
    if (self.archivedIds == nil) self.archivedIds = @[].mutableCopy;
    NSMutableDictionary *extendedRequest = @{@"type":@2, @"_id[nin]": self.archivedIds,
                                             @"sort_desc":@"last_message_sent_at",
                                             @"limit":@999}.mutableCopy;
    if ([ServicesManager instance].lastActivityDate != nil && self.dialogs.count > 0) {
        [extendedRequest setValue:@(ServicesManager.instance.lastActivityDate.timeIntervalSince1970)
                           forKey:@"updated_at[gte]"];
    } return extendedRequest;
}
```

**Comment by Pro100Andrey** [comment]
@iurichiba page limit - max 100

**Comment by iurichiba** [comment]
@Pro100Andrey that's it? Would that cause the crashes?
Or is it just an optimization issue? And what if I wanna get all of them (even if above 100)?

**Comment by Pro100Andrey** [comment]
@iurichiba max page limit for all condition : 100  this is a server constraints

**Comment by Pro100Andrey** [comment]
@iurichiba but [[ServicesManager instance].chatService allDialogsWithPa.. loaded recursive all items

**Comment by Pro100Andrey** [comment]
@iurichiba  Plus use` __weak __typeof(self)weakSelf = self;` for example:

```objc
- (void)loadRecentDialogsWithSuccess:(void(^)(NSArray *dialogObjects))completion
                          andFailure:(void(^)(QBResponse *response))failure{
 __weak __typeof(self)weakSelf = self;
    NSDictionary *request = [self requestForRecentDialogs];
    [[ServicesManager instance].chatService allDialogsWithPageLimit:999 extendedRequest:request iterationBlock:^(QBResponse *response, NSArray *dialogObjects, NSSet *dialogsUsersIDs, BOOL *stop) {
        weakSelf.recentDialogs = weakSelf.dialogs.mutableCopy;
        if (completion != nil) completion(dialogObjects);
    } completion:^(QBResponse *response) {
        if ([ServicesManager instance].isAuthorized) {
            if (response.success) [ServicesManager instance].lastActivityDate = [NSDate date];
        } if (failure != nil) failure(response);
    }];
}
```

**Comment by Pro100Andrey** [comment]
@iurichiba Please check this suggestion

**Comment by iurichiba** [comment]
@Pro100Andrey I'm gonna try your weakSelf suggestion. But should I change 999 to 100, too? I read the code, and it should work since you expect that behavior inside allDialogsWithPageLimit -- as you told us, it recursively gets all entries, I didn't understand why that would be a problem.

**Comment by Pro100Andrey** [comment]
@iurichiba >>>But should I change 999 to 100, too?  - yes, PageLimit need for  iteration

**Comment by iurichiba** [comment]
@Pro100Andrey but if I send 100, if someone has 250 dialogs for example, I wouldn't get the remaining 150, right? Should **I** implement pagination or iterate allDialogs again and again to get all 250?

Btw, why would that weakSelf solution fix EXC_BAD_ACCESS? Isn't there any danger that weakSelf is deallocated before iterationBlock runs? Should I assign it to a strong variable once iterationBlock begins like you use inside allDialogsWithPageLimit?

```obj-c
 __weak __typeof(self)weakSelf = self;
    __block void(^t_request)(QBResponsePage *responsePage);
    void(^request)(QBResponsePage *responsePage) = ^(QBResponsePage *responsePage) {
        
        __typeof(weakSelf)strongSelf = weakSelf;
        
        if (![strongSelf.serviceManager isAuthorized]) {
            if (completion) {
                completion(nil);
            }
            return;
        }

[...]
```

**Comment by Pro100Andrey** [comment]
@iurichiba 

    ```objc
[[ServicesManager instance].chatService allDialogsWithPageLimit:100 extendedRequest:request
                                                     iterationBlock:^(QBResponse *response, NSArray *dialogObjects, NSSet *dialogsUsersIDs, BOOL *stop)
    {
        //Add loaded page 
        [result addObjectsFromArray:dialogObjects];
    
    } completion:^(QBResponse *response) {
        //Loaded all pages
        weakSelf.recentDialogs = result.mutableCopy;
        
        if ([ServicesManager instance].isAuthorized) {
            if (response.success)
                [ServicesManager instance].lastActivityDate = [NSDate date];
            
        } if (failure != nil) failure(response);
    }];
```

**Comment by Pro100Andrey** [comment]
@iurichiba is this a full stack trace or there is more? [Most frequent (block_invoke, EXC_BAD_ACCESS): 93 crashes, 72 users, in 4 days](http://imgur.com/a/nEiyt)  ?

**Comment by Pro100Andrey** [comment]
@iurichiba please provide code snippet for 

```objc
[self loadRecentDialogsWithSuccess:^(NSArray *dialogObjects) {
      ?
       ....
    } andFailure:^(QBResponse *response) {
    ?
        ...
    }];
```

**Comment by iurichiba** [comment]
@Pro100Andrey here you go:
```obj-c
[self loadRecentDialogsWithSuccess:^(NSArray * _Nullable dialogObjects) {
                        [self updateDialogs];
                    } andFailure:nil];
```

`updateDialogs` separates the dialogs between recents and archived (we implemented an archiving mechanism) and sort them.

Yesterday, we analyzed our code and made some changes to it: we removed the iterationBlock (moving the code to the completionBlock) and implemented some minor improvements to it. Also, following your suggestions, we implemented the weakSelf solution.

We didn't deploy it, yet, but it looks like this, now:
```obj-c
- (void)loadRecentDialogsWithSuccess:(void(^)(NSArray *dialogObjects))completion
                          andFailure:(void(^)(QBResponse *response))failure {
    __weak typeof(self)weakSelf = self;
    NSDictionary *request = [self requestForRecentDialogs];
    [[ServicesManager instance].chatService allDialogsWithPageLimit:100 extendedRequest:request iterationBlock:nil completion:^(QBResponse *response) {
        weakSelf.recentDialogs = weakSelf.dialogs.mutableCopy;
        if (completion != nil) completion(weakSelf.dialogs);
        if ([ServicesManager instance].isAuthorized) {
            if (response.success) [ServicesManager instance].lastActivityDate = [NSDate date];
        } if (!response.success && failure != nil) failure(response);
    }];
}
```

We're gonna deploy it later today, and I think we can report the results back to you soon.
If you have any other insights, we'd love to hear them and make some more adjustments ASAP.

**Comment by VitaliyGuru** [comment]
@iurichiba Have you gotten any results?

**Comment by iurichiba** [comment]
@VitaliyGuru not yet, sorry. I submitted our update on friday, but Apple just approved it yesterday.
We're gonna let it run today, and check the results by the end of the day :)

**Comment by VitaliyGuru** [comment]
@iurichiba OK =) Looking forward to hearing from you.

**Comment by iurichiba** [comment]
Hi, @VitaliyGuru and @Pro100Andrey.

Unfortunately, it didn't work, we're still getting the same crashes... I'm out of ideas, I don't know what else could be wrong. Do you have any other insights?

Btw, we're migrating to an enterprise plan (due to shared instance instability), could that help? Also, using the enterprise plan, could we direct support so that we can look at the code together and check what could be wrong?

**Comment by Raikerian** [comment]
@iurichiba contact our support team directly and we can look into it together. Try to attach as much logs or actual cases as possible.

**Comment by iurichiba** [comment]
@Raikerian thanks, we'll contact the support team in the next few days.

**Comment by iurichiba** [comment]
Hello, I have an update on this issue. We managed to find out what caused it: the session wasn't being updated automatically (even though the documentation said that the SDK would handle that).

Now, we're checking if the session token is valid and if the token expiration date has passed before continuing the process. If it fails, we generate a new session and then try again.

Our crash-free skyrocketed after that. It's way more stable, now.

You should check it, I think there's something wrong in the SDK when it was supposed to renew the session -- hope you can find out what's wrong with that info!

I'm gonna reopen the issue just so that you can see it, you can close it afterwards :)

**Comment by Raikerian** [comment]
What exactly should we check? Any specific case on this?

**Comment by iurichiba** [comment]
I don't really know **what** you should check, but the case is:

- The user opens the app, and logs in (the session is created)
- The user puts the app in background
- A couple hours later, the user reopens the app, and it crashes because they tried to load his/her QBDialogs (using a QBRequest) but the session had expired

The SDK should renew the session before requesting the dialogs, shouldn't it?
> Stating from QuickBlox iOS SDK 2.4 you don't need to manage session manually. SDK will automatically do all this job.
>
> After login a REST session will be available for 2 hours.
>
> When the session expires, any QBRequest method will firstly renew it and then execute itself.
https://quickblox.com/developers/IOS#Overview

**Comment by Raikerian** [comment]
What SDK version are you using?

**Comment by iurichiba** [comment]
v2.9, the latest one

**Comment by Raikerian** [comment]
Our testers did not confirm this, as our project Q-municate and samples does not have this issue.
We would need some more details, maybe even test project if possible.,

You can also try and reproduce this in our samples by adding appropriate code lines. Until then we won't be able to help you with this as our projects does not have this issue.

**Comment by iurichiba** [comment]
Our project is working fine, now, since we're manually renewing the session when needed. I just reopened the issue so that I could tell you what happened with us in this case, and possibly help you find an issue with the SDK.

I can't confirm if there *is* a problem, I can only tell you what happened with us and how we fixed it, and that's why I reopened the card.

But since you couldn't find any problems, I think we can close the issue. There's no need to keep looking into this problem if we were the only ones experiencing it.

Thanks for the attention, either way :)
